fix: CalDAV credential verification between multiple instances

zomars opened the pull request:
_PR description is being written. Please check back in a minute._ 

Devin Session: https://app.devin.ai/sessions/bc70d3612a0b43778d98c30cf42d445c
    
<!-- This is an auto-generated description by cubic. -->
---

## Summary by cubic
Fixed a security issue where CalDAV credentials could be leaked between different server instances by ensuring credentials are only used with matching CalDAV servers.

- **Bug Fixes**
  - Added server URL checks to prevent using CalDAV credentials with the wrong server.
  - Improved logging for credential validation failures.

<!-- End of auto-generated description by cubic. -->



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Improved filtering of CalDAV calendars to ensure only calendars from the correct server are accessible with a given credential.
  * Enhanced validation and logging for CalDAV credential and calendar matching during event creation.

* **Bug Fixes**
  * Prevented potential credential leaks by ensuring CalDAV credentials are only matched with calendars from the same server.

* **Tests**
  * Added comprehensive tests for CalDAV credential validation, server URL extraction, and credential-to-calendar matching logic.
  * Expanded test coverage to verify prevention of CalDAV credential leaks and correct filtering behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

devin-ai-integration[bot] committed:
feat: prevent CalDAV credential leak between multiple instances

- Add server URL extraction from encrypted CalDAV credentials
- Add destination calendar server URL parsing
- Implement credential-to-destination server URL validation
- Filter CalDAV credentials by matching server URLs in createAllCalendarEvents
- Add comprehensive logging for credential validation failures
- Maintain backward compatibility for non-CalDAV integrations

Fixes security vulnerability where multiple CalDAV instances would
send credentials from first server to subsequent CalDAV servers.

Co-Authored-By: zomars@cal.com <zomars@me.com>

devin-ai-integration[bot] committed:
test: add comprehensive unit tests for CalDAV credential validation

- Test extractServerUrlFromCredential with valid/invalid credentials
- Test extractServerUrlFromDestination with various URL formats
- Test credentialMatchesDestination with matching/non-matching servers
- Cover error handling for decryption failures and invalid URLs
- Ensure fail-safe behavior prevents credential leakage

Co-Authored-By: zomars@cal.com <zomars@me.com>

devin-ai-integration[bot] committed:
fix: use environment variable for encryption key in tests

- Fix CI test failure by using process.env.CALENDSO_ENCRYPTION_KEY
- Ensures tests work in both local and CI environments
- All 22 unit tests continue to pass

Co-Authored-By: zomars@cal.com <zomars@me.com>

zomars committed:
Merge branch 'main' into devin/1751994162-fix-caldav-credential-leak

hbjORbj reviewed (commented):
1 non-block nit
1 Quesiton

pumfleet committed:
Merge branch 'main' into devin/1751994162-fix-caldav-credential-leak

zomars committed:
Refactor CalDAV calendar type to constant

Replaces hardcoded 'caldav_calendar' strings with a CALDAV_CALENDAR_TYPE constant throughout EventManager.ts for improved maintainability and consistency.

zomars commented on packages/lib/EventManager.ts:
```diff
@@ -172,6 +174,82 @@ export default class EventManager {
 
+  private extractServerUrlFromCredential(credential: CredentialForCalendarService): string | null {
+    try {
+      if (credential.type !== "caldav_calendar") {
```
addressed

zomars commented:
feedback addressed @calcom/foundation

alishaz-polymath committed:
fix credential leak

alishaz-polymath committed:
early return

alishaz-polymath committed:
Merge branch 'main' into devin/1751994162-fix-caldav-credential-leak

alishaz-polymath committed:
add test

alishaz-polymath commented:
Test Added, ready for review @emrysal 🙏

alishaz-polymath reviewed (dismissed):
Tested locally and it works. LGTM

emrysal committed:
More idiomatic mock, tiny nit

emrysal committed:
Merge branch 'main' into devin/1751994162-fix-caldav-credential-leak

emrysal reviewed (approved):
LGTM 🚀
